src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java

```diff
@@ -315,9 +319,39 @@ public GroupShardsIterator groupByAllIt() {
         return new GroupShardsIterator(set);
     }
 
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        IndexRoutingTable that = (IndexRoutingTable) o;
+
+        if (!index.equals(that.index)) return false;
+        if (!shards.equals(that.shards)) return false;
+
+        return true;
+    }
+
+    @Override
+    public int hashCode() {
+        int result = index.hashCode();
+        result = 31 * result + shards.hashCode();
+        return result;
+    }
+
     public void validate() throws RoutingValidationException {
     }
 
+    @Override
+    public IndexRoutingTable readFrom(StreamInput in) throws IOException {
+        return Builder.readFrom(in);
```

Comment:
same question as above why is the read / wrtie code on the Builder? seems wrong, I would have expected it here?